packages/core/src/auth/auth.ts

```diff
@@ -890,15 +891,81 @@ export class Auth implements AuthService, ConnectionManager {
             : `${localizedText.iamIdentityCenter} (${truncatedUrl})`
     }
 
-    public async createConnectionFromProfile(connection: Connection, profile: SsoProfile) {
-        const id = uuid.v4()
-        getLogger().info(`Creating new connection ${id} from existing connection ${connection.id}`)
+    // Used by Amazon Q to re-use connection from AWS Toolkit listConnection API response
+    public async createConnectionFromApi(connection: AwsConnection) {
+        getLogger().info(`Reusing connection ${connection.id}`)
+        const profile = {
+            type: connection.type,
+            ssoRegion: connection.ssoRegion,
+            scopes: connection.scopes,
+            startUrl: connection.startUrl,
+        } as SsoProfile
+        const id = connection.id
         const storedProfile = await this.store.addProfile(id, profile)
-        await this.updateConnectionState(id, 'valid')
-        const provider = this.getSsoTokenProvider(id, storedProfile)
-        await provider.duplicateToken(connection.id, id)
+        await this.updateConnectionState(id, connection.state)
         return this.getSsoConnection(id, storedProfile)
     }
+
+    // Used by AWS Toolkit to update connection status & scope when this connection is updated by Amazon Q
+    // If such connection does not exist, create one with same id.
+    // Otherwise, update its scope and/or state.
+    public async updateConnectionFromApi(connection: AwsConnection) {
+        getLogger().info(`Updating connection ${connection.id} from API`)
+        const id = connection.id
+        const profile = this.store.getProfile(connection.id)
+        const newProfile = {
+            type: connection.type,
+            ssoRegion: connection.ssoRegion,
+            scopes: connection.scopes,
+            startUrl: connection.startUrl,
+        } as SsoProfile
+        if (profile === undefined) {
+            await this.store.addProfile(id, newProfile)
```

Comment:
so update is also upsert here? if Q adds a new connection, it will be registered in the toolkit?

if so, can we rename this function to upsert to be more clear

Reply:
could name it "set" instead of "update". Reason I suggested "update" is because that's the convention used by `memento` / globalStorage

Reply:
"set" is good

Reply:
Renamed into `setConnections`